apisix/balancer/ewma.lua

```diff
@@ -0,0 +1,237 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+
+local resty_lock = require("resty.lock")
+local core = require("apisix.core")
+
+local _M = {}
+local DECAY_TIME = 10 -- this value is in seconds
+local LOCK_KEY = ":ewma_key"
+local PICK_SET_SIZE = 2
+local ewma_lock, ewma_lock_err 
+
+
+local function lock(upstream)
+	local _, err = ewma_lock:lock(upstream .. LOCK_KEY)
+	if err then
+		if err ~= "timeout" then
+			core.log.error(string.format("EWMA Balancer failed to lock: %s", tostring(err)))
+		end
+	end
+
+	return err
+end
+
+local function unlock()
+	local ok, err = ewma_lock:unlock()
+	if not ok then
+		core.log.error(string.format("EWMA Balancer failed to unlock: %s", tostring(err)))
+	end
+
+	return err
+end
+
+local function decay_ewma(ewma, last_touched_at, rtt, now)
+	local td = now - last_touched_at
+	td = (td > 0) and td or 0
+	local weight = math.exp(-td/DECAY_TIME)
+
+	ewma = ewma * weight + rtt * (1.0 - weight)
+	return ewma
+end
+
+local function store_stats(upstream, ewma, now)
+	local success, err, forcible = ngx.shared.balancer_ewma_last_touched_at:set(upstream, now)
+	if not success then
+		core.log.warn("balancer_ewma_last_touched_at:set failed " .. err)
```

Comment:
bad performance: avoid to create new string object, this way will make GC work hard.